Common Threats: Supply chain attacks (#2467)
privacyguides/privacyguides.org: docs/basics/common-threats.md

If you are concerned about **physical attacks** you should use an operating system with a secure verified boot implementation, such as Android, iOS, macOS, or [Windows (with TPM)](https://learn.microsoft.com/windows/security/information-protection/secure-the-windows-10-boot-process). You should also make sure that your drive is encrypted, and that the operating system uses a TPM or Secure [Enclave](https://support.apple.com/guide/security/secure-enclave-sec59b0b31ff/1/web/1) or [Element](https://developers.google.com/android/security/android-ready-se) to rate limit attempts to enter the encryption passphrase. You should avoid sharing your computer with people you don't trust, because most desktop operating systems don't encrypt data separately per-user.

> <span class="pg-amber">:material-package-variant-closed-remove: Supply Chain</span>

Reviewer: I'm wondering if, instead of placing this here, we add a new section at the bottom of the page: "Advanced Persistent Threats" where we cover larger-scale issues like this. Thoughts?

Author: I kind of like that idea.

Author: I'm wondering whether this actually is a good idea at all.

Supply chain attacks don't necessarily have to come from an APT. For example in the case of naming similar nodejs libs that are often typo'ed or uploading some crappy crypto wallet to a store, these are not done by an APT.

> Supply chain attacks are a type of <span class="pg-orange">:material-bug-outline: Passive Attack</span> which can be part of a larger <span class="pg-red">:material-target-account: Targeted Attack</span>.

Reviewer: I think there are many more cases of these being targeted attacks than the other way round.

```suggestion
Supply chain attacks are frequently a form of <span class="pg-red">:material-target-account: Targeted Attack</span> towards businesses, governments, and activists, although they can end up compromising the public at large as well.

<div class="admonition example" markdown>
<p class="admonition-title">Example</p>

A notable example of this occurred in 2017 when M.E.Doc, a popular accounting software in Ukraine, was infected with the *NotPetya* virus, subsequently infecting people who downloaded that software with ransomware. NotPetya itself was a ransomware attack which impacted 2000+ companies in various countries, and was based on the *EternalBlue* exploit developed by the NSA to attack Windows computers over the network.

</div>
```

Author: > these being targeted attacks than the other way round

I guess it depends on how you look at it. Targeted in respect to a library, or piece of software. Passive in the definition that it's supposed to target a lot of people.

But now I think about it I like your definition better.

There are few ways in which this type of attack might be carried out:

> 1. A contributor or employee might work their way into a position of power within an organization and then abuse that position to add some malicious code.

Reviewer: ```suggestion
1. A contributor or employee might work their way into a position of power within a project or organization, then abuse that position by adding malicious code.
```

2. A developer may be coerced by an outside party to add malicious code.
3. An individual or group might identify a third party software dependency (also known as a library) and work to infiltrate it with the above two methods, knowing that it will be used by "downstream" software developers.

These sorts of attacks can require a lot of time and preparation to perform and are risky because they can be detected, particularly in open source projects if they are popular and have outside interest. Unfortunately they're also one of the most dangerous as they can never be entirely mitigated. We would encourage readers only use software which has a good reputation and makes an effort to reduce risk by:

1. Only adopting popular software that has been around for a while. The more interest in a project the, the greater likelihood that external parties will notice malicious changes. A malicious actor will also need to spend more time gaining community trust with meaningful contributions.

> 2. Automated build infrastructure for released binary code reduces the attack surface such as, a compromised developer workstation or breached server. It provides accountability to others involved in the project in that binaries are produced from a central source code repository.

Reviewer 2: Is this recommending automated build infrastructure over local developer workstations / breached servers? If so, I think it could be rewritten better:
```suggestion
2. Make sure software you use builds released binaries with widely-used, trusted build infrastructure platforms (i.e. GitHub workflows) as opposed to developer workstations or self-hosted servers. This is in order to reduce the attack surface and give confidence that the binaries produced are in fact produced correctly.
```
Imo, the best strategy is reproducible builds cross-checked from multiple parties (more reason to adopt nixos lol) but I think relying on centralized build infrastructure is also a good strategy for people without the time or compute.

Reviewer: This advice is interesting, because it is advocating for the exact thing which hid the `xz` problem 😅 

I mean I suppose not all vulnerabilities are going to be on that level of sophistication, but right now this list doesn't mention compiling from source yourself at all, and it probably should.

Author: > This advice is interesting, because it is advocating for the exact thing which hid the `xz` problem 😅

Its supposed to read as the opposite to that. Ideally tarballs aren't offered only tagged releases that come from a git source.

Author: Yes, going to approve this because we do value [reproducible builds](https://reproducible-builds.org/) elsewhere on the site.

3. Code signing on individual commits and releases increases an auditable trail of who did what. For example: Was the malicious code in the software repository? Which developer added it? Was it added during the build process?

> 4. In open source projects the code should have commit messages which are explain exactly what the associated code does. This makes it easier for external parties to verify and detect if the code doesn't match the description.

Reviewer 2: Not sure how relevant commit messages are. They are useful for devs for git hygiene, but for security, I imagine any external party is gonna read the code anyway. (poor commit messages don't necessarily mean bad security). 

Maybe I'm just projecting tho :sweat_smile:

Author: That's the thing they've won't. I've noticed a clear pattern where projects with a heap of commits that just say "fix" basically get no external attention. Nobody really cares about those projects.

Author: I reworded this to "audit", because we do use that word a lot.

The easier it is for outsiders to get an understanding of how things are implemented better. This might also mean it's easier to track a regression that is actually a cleverly designed exploit.

> 5. The number of contributors or maintainers, a lone developer may be more
> susceptible to being coerced or forced into adding malicious code by an external party.
> 6. Economical pressure that would occur if a company was involved with negligent behavior that enabled malicious code to be added. This may very well mean software developed by "Big Tech" has more scrutiny than a lone developer who doesn't answer to anyone.

Reviewer: ```suggestion
5. The number of contributors or maintainers, a lone developer may be more susceptible to being coerced or forced into adding malicious code by an external party.
6. Economical pressure that would occur if a company was involved with negligent behavior that enabled malicious code to be added. This may very well mean software developed by "Big Tech" has more scrutiny than a lone developer who doesn't answer to anyone.
```

So... do these say the same thing though? Feels like we could go like this instead:

```suggestion
5. Noting the number of contributors or maintainers a program has. A lone developer may be more susceptible to being coerced into adding malicious code by an external party, or to negligently enable undesirable behavior.
```

Reviewer 3: I feel as though it could be worth keeping this sentence:

> This may very well mean software developed by "Big Tech" has more scrutiny than a lone developer who doesn't answer to anyone.

Otherwise the combined paragraph is better.

Author: I do as well, the reason being is because we see a lot of people thinking they are increasing privacy by shifting to a small service that may not be very mature, instead of one which is well developed, or resourced or even has E2EE.

Author: Applied what both @freddy-m and @jonaharagon wrote.

If you are concerned about **physical attacks** you should use an operating system with a secure verified boot implementation, such as Android, iOS, macOS, or [Windows (with TPM)](https://learn.microsoft.com/windows/security/information-protection/secure-the-windows-10-boot-process). You should also make sure that your drive is encrypted, and that the operating system uses a TPM or Secure [Enclave](https://support.apple.com/guide/security/secure-enclave-sec59b0b31ff/1/web/1) or [Element](https://developers.google.com/android/security/android-ready-se) to rate limit attempts to enter the encryption passphrase. You should avoid sharing your computer with people you don't trust, because most desktop operating systems don't encrypt data separately per-user.

> <span class="pg-viridian">:material-package-variant-closed-remove: Supply Chain</span>

Reviewer: ```suggestion
<span class="pg-viridian">:material-package-variant-closed-remove: Supply Chain Attacks</span>
```

</div>

There are few ways in which this type of attack might be carried out:

1. A contributor or employee might work their way into a position of power within a project or organization, then abuse that position by adding malicious code.
2. A developer may be coerced by an outside party to add malicious code.
3. An individual or group might identify a third party software dependency (also known as a library) and work to infiltrate it with the above two methods, knowing that it will be used by "downstream" software developers.

These sorts of attacks can require a lot of time and preparation to perform and are risky because they can be detected, particularly in open source projects if they are popular and have outside interest. Unfortunately they're also one of the most dangerous as they are very hard to mitigate entirely. We would encourage readers only use software which has a good reputation and makes an effort to reduce risk by:

> 1. Only adopting popular software that has been around for a while. The more interest in a project the, the greater likelihood that external parties will notice malicious changes. A malicious actor will also need to spend more time gaining community trust with meaningful contributions.

Reviewer: ```suggestion
1. Only adopting popular software that has been around for a while. The more interest in a project the greater likelihood that external parties will notice malicious changes. A malicious actor will also need to spend more time gaining community trust with meaningful contributions.
```

> 2. Make sure software you use builds released binaries with widely-used, trusted build infrastructure platforms (i.e. GitHub workflows) as opposed to developer workstations or self-hosted servers. This is in order to reduce the attack surface and give confidence that the binaries produced are in fact produced correctly.

Reviewer: ```suggestion
2. Finding software which releases binaries with widely-used, trusted build infrastructure platforms, as opposed to developer workstations or self-hosted servers. Some systems like GitHub Actions let you inspect the build script that runs publicly for extra confidence. This lessens the likelihood that malware on a developer's machine could infect their packages, and gives confidence that the binaries produced are in fact produced correctly.
```

i think this is worded weird; also added info on why workflows are cool

> 3. Code signing on individual commits and releases increases an auditable trail of who did what. For example: Was the malicious code in the software repository? Which developer added it? Was it added during the build process?

Reviewer: ```suggestion
3. Looking for code signing on individual source code commits and releases, which creates an auditable trail of who did what. For example: Was the malicious code in the software repository? Which developer added it? Was it added during the build process?
```

All of these bullet points should be worded as if they finish the sentence "...and makes an effort to reduce risk by:" because that is how the list begins. Like if you read it out loud, "reduce risk by looking for code signing..." etc.

Author: I do agree on that, i think i had bullet points originally but couldn't figure out how to indent it properly to be a sublist.

> 4. In open source projects the code should have meaningful commit messages (such as [conventional commits](https://conventionalcommits.org)) which explain what the change is supposed accomplish. Clear messages make it easier for outsiders to the project to verify, audit and find bugs.

Reviewer: ```suggestion
4. Checking whether the source code has meaningful commit messages (such as [conventional commits](https://conventionalcommits.org)) which explain what the change is supposed to accomplish. Clear messages can make it easier for outsiders to the project to verify, audit, and find bugs.
```
